ndarray-rand/src/lib.rs

```diff
@@ -280,6 +358,41 @@ where
         };
         self.select(axis, &indices)
     }
+
+    fn shuffle_axis_inplace(&mut self, axis: Axis)
+    where
+        D: RemoveAxis,
+        S: DataMut,
+    {
+        self.shuffle_axis_inplace_using(axis, &mut get_rng())
+    }
+
+    fn shuffle_axis_inplace_using<R>(&mut self, axis: Axis, rng: &mut R)
+    where
+        R: Rng + ?Sized,
+        D: RemoveAxis,
+        S: DataMut,
+    {
+        for i in (1..self.len_of(axis)).rev() {
+            // Invariant: elements with index > i have been locked in place.
+            let j = rng.gen_range(0, i + 1);
+
+            if i != j {
+                // Swap the two slices along `axis`
+                let slice1 = self.index_axis(axis, i);
+                let slice2 = self.index_axis(axis, j);
+
+                for (x, y) in slice1.iter().zip(slice2.iter()) {
```

Comment:
We are modifying through read-only/shared array views and this is invalid, unless we have interior mutability (i.e using Cell, Mutex or similar).

And we'd like to avoid `iter().zip()` for ndarray, because it is much slower than it needs to be, we have Zip.

Reply:
I'm working on making `ArrayView<Cell<T>, D>` a reality - it allows us to have.. more fun like this. :)

Reply:
> We are modifying through read-only/shared array views and this is invalid, unless we have interior mutability (i.e using Cell, Mutex or similar).

On the other hand the method is taking a mutable reference to `self`, hence we are guaranteed to be the only ones operating on the array. Or am I interpreting this incorrectly?

Reply:
Modifying through an ArrayView (without the Cell) exception is full stop invalid, because we treat it as we would treat a `&[T]`. (We could go into details and exceptions at another time, maybe there's some argument for a more lenient view?)

Reply:
In the short range it is easy to point to what is invalid: `x` is a `&T` but we are modifying through it. That's not permitted.

Exceptions would come in to place if you somehow got a raw pointer out of an array view, *maybe* you'd then be able to make an argument. But since we go through explicit shared references here, we are at invalid full stop.

Reply:
Given the `S: DataMut` trait bound on the method, aren't we dealing with an `ArrayViewMut`, which is equivalent to a `&mut [T]`? 
Just trying to *truly* understand, I am not trying to argue :sweat_smile:

Reply:
Sure, no problem. The types matter and there are shared refs used on the path to the swap. If we read the reference this is UB, we can't just cast from &T and mutate.